### ISSUE TITLE
Field Validation Priority Order

### ISSUE THREAD

**Issue by rodleviton** [odecee/angular-form-lib#6]
Hi Brett,

I am noticing that if fields are inside of nested forms or wrapped in ng-if conditionals when the form is submitted they are pushed to the bottom of the field validation order. This is causing some interesting side effects when coupled with the duScroll library. It would be great to always have fields validated in the correct DOM order that they are in.

Cheers
Rod

**Comment by rodleviton** [comment]
Potential solution could be to evaluate first error field in form being submitted via '.ng-invalid' class.

```
applyBehaviour: function(fieldElem, fieldState, formSubmitAttempted) {

  // Set the focus to the field if there is an error showing and a form-submit has been attempted
  if (fieldState === 'error' && formSubmitAttempted) {

    // Make sure element is the first field with an error based on DOM order
    var firstElement = $document[0][focusController.$name].querySelectorAll('.ng-invalid')[0];
    var isFirstElement = (firstElement.id === fieldElem[0].id);

    if (!focusController._focusErrorElement && isFirstElement && setFocusOnField($document, $timeout, duScrollDuration, fieldElem, formController._policy.fieldFocusScrollOffset)) {
      focusController._focusErrorElement = fieldElem;
    }
  }
},
```

**Comment by uglow** [comment]
Great suggestion, Rod. Does `querySelectorAll()` make any guarantees about the order of results vs DOM order?

**Comment by rodleviton** [comment]
Hi Brett

According to the [spec](https://www.w3.org/TR/selectors-api/#queryselectorall) it returns a nodeList in document order.

**Comment by uglow** [comment]
Great, thanks. I'll try to get this out in the next couple of days.

**Comment by rodleviton** [comment]
Small update. Probably want to check if there is a firstElement before running an comparisons.

var isFirstElement = firstElement ? (firstElement.id === fieldElem[0].id) : false;

**Comment by uglow** [comment]
Hi Rod,

This is harder to achieve because the fields are added and removed to NgFormModelController, which has no idea about the DOM order. 

Not sure what to do?

**Comment by uglow** [comment]
Fixed in be1b47df5e1574010232e3a0ba3bfd0f9493d450 (actually fixed in df6125af1a0368cc36584b3e78ff5a94301a0cbe, but needed to update Travis config once-or-ten-times to get everything automated).

See the [dynamic form demo](https://odecee.github.io/angular-form-lib/#/formDemos#formDemo5) for proof of fix. Thanks for your help @rodleviton  👍
